### DISCUSSION TITLE
Fastapi

### DISCUSSION THREAD

**Discussion by kumar045** [fastapi/fastapi#6093] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
import gradio as gr

CUSTOM_PATH = "/gradio"

app = FastAPI()

@app.get("/")
def read_main():
    return {"message": "This is your main app"}

io = gr.Interface(lambda x: "Hello, " + x + "!", "textbox", "textbox")
gradio_app = gr.routes.App.create_app(io)

app.mount(CUSTOM_PATH, gradio_app)
```

### Description

I want to restrict /gradio that only login user can access this 
Can I use some condition like user login and subscribed or not 

### Operating System

Windows

### FastAPI Version

Latest install with pip

### Python Version

3.9

**Comment by moadennagi** [comment]
I would use a middleware and plug it when mounting gradio app to fastapi app. Here is an example using starlette.middleware.base

```python
from starlette.middleware import Middleware
from starlette.middleware.base import BaseHTTPMiddleware

from starlette.routing import Mount
from fastapi import FastAPI, HTTPException

class CustomHeaderMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next):
        auth_header = request.headers.get('Authorization')
        if not auth_header:
            raise HTTPException(401)
        # Logic for parsing token, getting user ...
        response = await call_next(request)
        return response

subapi = FastAPI()

@subapi.get("/sub/1")
def read_sub():
    return {"message": "Hello World from sub API / 1"}

subapi.get("/sub/2")
def sub2():
    return {"message": "Hello World from sub API / 2"}


middleware = [
    Middleware(CustomHeaderMiddleware)
]

routes = [
    Mount('/subapi', app=subapi, middleware=middleware),
]

app = FastAPI(routes=routes)

@app.get("/app")
def read_main():
    return {"message": "Hello World from main app"}
```

**Comment by moadennagi** [comment, marked as answer]
I looked into gradio documentation (https://gradio.app/sharing_your_app/#accessing-the-network-request-directly), gradio provides ability to access the request data on the main function, you can access session cookies, headers ... and do various checks, here is a simplified example:

```python
import gradio as gr
from fastapi import FastAPI, HTTPException, Response

CUSTOM_PATH = "/gradio"

app = FastAPI()

@app.get("/")
def read_main():
    return {"message": "This is your main app"}

# I added this as a get request to be able to easily set cookie through the browser 
@app.get('/login')
def dummy_login(response: Response):
    response.set_cookie(key='fakesession', value='fake-cookie-session-value')
    return {'message': 'logged in'}

def echo(name, request: gr.Request):
    if request:
        try:
            getattr(request.cookies, 'fakesession')
            # do some checks and validations ...
        except AttributeError:
            raise HTTPException(401, detail='Must be logged in')
        return "Hello, " + name + "!"
    return name

io = gr.Interface(echo, "textbox", "textbox")
app = gr.mount_gradio_app(app, io, path=CUSTOM_PATH)

```
This other example uses middleware to run before all requests to gradio path:
```python
import gradio as gr

from fastapi import FastAPI, Request

CUSTOM_PATH = "/gradio"

app = FastAPI()

@app.get("/")
def read_main():
    return {"message": "This is your main app"}

@app.middleware('http')
async def auth(request: Request, call_next):
    if request.scope['path'].startswith(CUSTOM_PATH):
        # do some auth checks ...
        print('checks')
        pass
    response = await call_next(request)
    return response

io = gr.Interface(lambda x: "hi" + x, "textbox", "textbox")
app = gr.mount_gradio_app(app, io, path=CUSTOM_PATH)
```
I'm not sure using `startswith` is the best option as it would cover all paths starting with gradio (gradio/... but also gradiosometext), you could me refine that using regex.

**Comment by kumar045** [comment]
Thank you very very much it worked 
Other example I tried and it worked
